tuf/repository_tool.py

```diff
@@ -3100,50 +3102,44 @@ def load_repository(repository_directory, repository_name='default',
   repository, consistent_snapshot = repo_lib._load_top_level_metadata(repository,
     filenames, repository_name)
 
-  # Load the delegated targets metadata and generate their fileinfo.  The
-  # extracted fileinfo is stored in the 'meta' field of the snapshot metadata
-  # object.
+  delegated_roles_filenames = repo_lib.get_delegated_roles_metadata_filenames(
+      metadata_directory, consistent_snapshot, storage_backend)
+
+  # Load the delegated targets metadata and their fileinfo.
+  # The delegated targets roles form a tree/graph which is traversed in a
+  # breadth-first-search manner starting from 'targets' in order to correctly
+  # load the delegations hierarchy.
   targets_objects = {}
-  loaded_metadata = []
   targets_objects['targets'] = repository.targets
 
-  metadata_files = sorted(storage_backend.list_folder(metadata_directory),
-      reverse=True)
-  for metadata_role in metadata_files:
-
-    metadata_path = os.path.join(metadata_directory, metadata_role)
-    metadata_name = \
-      metadata_path[len(metadata_directory):].lstrip(os.path.sep)
-
-    # Strip the version number if 'consistent_snapshot' is True,
-    # or if 'metadata_role' is Root.
-    # Example:  '10.django.json' --> 'django.json'
-    consistent_snapshot = \
-      metadata_role.endswith('root.json') or consistent_snapshot == True
-    metadata_name, junk = repo_lib._strip_version_number(metadata_name,
-      consistent_snapshot)
-
-    if metadata_name.endswith(METADATA_EXTENSION):
-      extension_length = len(METADATA_EXTENSION)
-      metadata_name = metadata_name[:-extension_length]
-
-    else:
-      logger.debug('Skipping file with unsupported metadata'
-          ' extension: ' + repr(metadata_path))
-      continue
-
-    # Skip top-level roles, only interested in delegated roles now that the
-    # top-level roles have already been loaded.
-    if metadata_name in ['root', 'snapshot', 'targets', 'timestamp']:
+  # A deque used to keep the next delegation to be loaded
+  delegations = deque()
```

Comment:
I'm curious about the use of a deque here, can you explain the rationale a bit?
AFAICT you're just using it as a FIFO, at which point a list with `pop(0)` would work just as well and would be a bit more familiar.

Reply:
Answering both of your comments @lukpueh @joshuagl :

Yes, @joshuagl got it right, the idea is to use a FIFO and continuously append the next delegations while removing(`pop(0)/popleft()`) the loaded one (unlike in my first implementation where the delegations were only appended and never removed from the list). 

@joshuagl my reasoning about [list vs deque as a FIFO](https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-queues) 

Using only a FIFO structure would be enough if there were no potential cycles in the graph, but since what we want to achieve is traverse the graph and avoid loops, we need one ordered structure + one fast lookup structure, hence the deque+set (or one that can do both which would be the dict in Python 3.x).

_And for a completeness of this discussion ..._ 
On the other hand, up to now we don't have exact constraints in `load_repository()` regarding the graph traversal  and we can as well use a [list as a stack (LIFO)](https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-stacks) which will result in a depth-first-search type of traversal.

I'm open to suggestions about less magical code or more pythonic way of implementing this :)

Reply:
> Yes, @joshuagl got it right, the idea is to use a FIFO and continuously append the next delegations while removing(pop(0)/popleft()) the loaded one (unlike in my first implementation where the delegations were only appended and never removed from the list).

Sure, but your first approach already implemented a FIFO queue, didn't it? The only difference that I saw is that the list grows in your first approach but does not in your second. 

```
# Go from left to right, adding to to the right
  ↓
| 0 | 
      ↓
| 0 | 1 |
          ↓
| 0 | 1 | 2 | 
```

```
# Pop from left, add to the right
  ↓
| 0 |
      ↓
    | 1 | 
          ↓
        | 2 |
```

Reply:
> hence the deque+set (or one that can do both which would be the dict in Python 3.x).

Haven't checked, but would `collections.OrderedDict` have the desired properties cross-version?

Reply:
> On the other hand, up to now we don't have exact constraints in load_repository() regarding the graph traversal and we can as well use a list as a stack (LIFO) which will result in a depth-first-search type of traversal.

AFAIU, when loading the delegation graph into memory it really shouldn't matter if we do depth or breadth first. This is only important when verifying targets.

Reply:
> > Yes, @joshuagl got it right, the idea is to use a FIFO and continuously append the next delegations while removing(pop(0)/popleft()) the loaded one (unlike in my first implementation where the delegations were only appended and never removed from the list).
> 
> Sure, but your first approach already implemented a FIFO queue, didn't it? The only difference that I saw is that the list grows in your first approach but does not in your second.
> 
> ```
> # Go from left to right, adding to the right
>   ↓
> | 0 | 
>       ↓
> | 0 | 1 |
>           ↓
> | 0 | 1 | 2 | 
> ```
> 
> ```
> # Pop from left, add to the right
>   ↓
> | 0 |
>       ↓
>     | 1 | 
>           ↓
>         | 2 |
> ```

Yes, yes, exactly :) 
 1) Yes, the goal here was to save memory 2) This could've been achieved with a list and pop(0) but this is a slow operation so here comes the deque "which was designed to have fast appends and pops from both ends"

Reply:
> > hence the deque+set (or one that can do both which would be the dict in Python 3.x).
> 
> Haven't checked, but would `collections.OrderedDict` have the desired properties cross-version?

Probably it would do the job, I think I had overlooked it because it sounded outdated from the perspective of the newer Python versions ...

Reply:
IMO really any of the approaches discussed here (FIFO or LIFO with list + set, deque + set or OrderedDict) is absolutely fine, I think I would even have been okay with not guarding against infinite loops, as long as it is documented. :)

So what do you think about keeping it as it is and adding another comment about saving memory by using a deque + leftpop?

Reply:
+ warning the user when a loop is encountered.

Reply:
> * warning the user when a loop is encountered.

Is this related to the sentence above (_"okay with not guarding against infinite loops"_) but appearing as a bullet? 😁 
Otherwise, yep, it works for me :)

Reply:
haha, this was meant to be a `+` as in:

*... what do you think  about adding another comment ... **plus**  warning the user when a loop is encountered?*